packages/synapse-react-client/src/components/SynapseTable/SynapseTable.tsx

```diff
@@ -1,889 +1,277 @@
-          }
-        })
-      } catch (err) {
-        console.error('Error on getGroupHeaders batch: ', err)
-      }
-    }
-    if (userProfileIds.length > 0) {
-      try {
-        const data = await getUserProfileWithProfilePicAttached(userProfileIds)
-        data.list.forEach((el: UserProfile) => {
-          mapUserIdToHeader[el.ownerId] = el
+  const isShowingAccessColumn: boolean = Boolean(
+    showAccessColumn &&
+      entity &&
+      isEntityViewOrDataset(entity) &&
+      allRowsHaveId(data),
+  )
+
+  const rowsAreDownloadable =
+    entity && isFileViewOrDataset(entity) && isLoggedIn && allRowsHaveId(data)
+
+  const isShowingDirectDownloadColumn = Boolean(
+    rowsAreDownloadable && showDirectDownloadColumn,
+  )
+
+  const isShowingAddToV2DownloadListColumn: boolean = Boolean(
+    rowsAreDownloadable &&
+      !hideAddToDownloadListColumn &&
+      !isRowSelectionVisible,
+  )
+
+  const columns: ColumnDef<Row, string | null>[] = useMemo(
+    () => [
+      rowSelectionColumn,
+      addToDownloadListColumn,
+      directDownloadColumn,
+      accessColumn,
+      ...(data?.columnModels?.map((columnModel, index) => {
+        return columnHelper.accessor(row => row.values[index], {
+          id: columnModel.name,
+          enableSorting: isSortableColumn(columnModel.columnType),
+          enableResizing: true,
+          header: TableDataColumnHeader,
+          cell: TableDataCell,
         })
-      } catch (err) {
-        console.error('Error on getUserProfile : ', err)
-      }
-    }
-    this.setState({
-      mapEntityIdToHeader,
-      mapUserIdToHeader,
-      isFetchingEntityHeaders: false,
-    })
-  }
+      }) ?? []),
+    ],
+    [data?.columnModels],
+  )
+
+  const prependColumnVisibility: VisibilityState = useMemo(
+    () => ({
+      [rowSelectionColumn.id as string]: isRowSelectionVisible,
+      [addToDownloadListColumn.id as string]:
+        isShowingAddToV2DownloadListColumn,
+      [directDownloadColumn.id as string]: isShowingDirectDownloadColumn,
+      [accessColumn.id as string]: isShowingAccessColumn,
+    }),
+    [
+      isRowSelectionVisible,
+      isShowingAccessColumn,
+      isShowingAddToV2DownloadListColumn,
+      isShowingDirectDownloadColumn,
+    ],
+  )
+
+  // Transform our `columnsToShowInTable` to @tanstack/react-table's `VisibilityState`
+  const columnVisibility: VisibilityState = useMemo(
+    () =>
+      (data?.columnModels ?? []).reduce((prev: VisibilityState, curr) => {
+        return {
+          ...prev,
+          [curr.name]: columnsToShowInTable.includes(curr.name),
+        }
+      }, prependColumnVisibility),
+    [columnsToShowInTable, data?.columnModels, prependColumnVisibility],
+  )
+
+  const table = useReactTable({
+    data: data?.queryResult?.queryResults.rows ?? [],
+    columns,
+    getCoreRowModel: getCoreRowModel(),
+    manualSorting: true,
+    manualFiltering: true,
+    onSortingChange: setSort,
+    enableMultiSort: true,
```

Comment:
In one design, I remember someone on the design team indicating that they wanted to remove the ability to sort multiple columns at once, but this doesn't need to be verified/implemented in this PR.